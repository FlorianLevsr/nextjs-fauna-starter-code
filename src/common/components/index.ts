export { default as AddTaskForm } from './add-task-form'
export { default as TaskList } from './task-list'
