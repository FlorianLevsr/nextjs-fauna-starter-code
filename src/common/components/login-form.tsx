import { FormControl, FormLabel } from '@chakra-ui/form-control'
import {
  Alert,
  AlertIcon,
  AlertTitle,
  AlertDescription,
  Input,
  Center,
  Button,
  InputProps,
} from '@chakra-ui/react'
import React, { FC, FormEventHandler, useState } from 'react'
import { MutationResult } from 'react-apollo'

interface LoginFormProps {
  callback: (username: string, password: string) => Promise<void>
  mutationResult: MutationResult<unknown>
}

type LoginFormInputType = 'username' | 'password'

type LoginFormData = Record<LoginFormInputType, string>

interface LoginFormInputProps extends InputProps {
  label: string
  inputName: LoginFormInputType
  setProperty: (inputName: LoginFormInputType, inputValue: string) => void
}

const LoginFormInput: FC<LoginFormInputProps> = ({
  label,
  inputName,
  setProperty,
  type,
  ...rest
}) => (
  <FormLabel>
    {label}
    <Input
      {...rest}
      type={type || 'text'}
      name={inputName}
      onChange={(event) => setProperty(inputName, event.target.value)}
    />
  </FormLabel>
)

const LoginForm: FC<LoginFormProps> = ({ callback, mutationResult }) => {
  const { error, loading } = mutationResult

  const [formData, setFormData] = useState<LoginFormData>({
    username: '',
    password: '',
  })

  const setProperty = (
    inputName: LoginFormInputType,
    inputValue: string
  ): void => {
    const newFormData = { ...formData, [inputName]: inputValue }
    setFormData(newFormData)
  }

  const onSubmitHandler: FormEventHandler = (event) => {
    event.preventDefault()
    const { username, password } = formData
    callback(username, password)
  }

  return (
    <form onSubmit={onSubmitHandler}>
      {error && (
        <Alert status="error" mr={4}>
          <AlertIcon />
          <AlertTitle mr={2}>Error:</AlertTitle>
          <AlertDescription>{error.message}</AlertDescription>
        </Alert>
      )}
      <FormControl>
        <LoginFormInput
          label="Username"
          inputName="username"
          setProperty={setProperty}
        />
        <LoginFormInput
          label="Password"
          inputName="password"
          type="password"
          setProperty={setProperty}
        />
        <Center>
          <Button
            isLoading={loading}
            type="submit"
            colorScheme="teal"
            onClick={onSubmitHandler}
          >
            Submit
          </Button>
        </Center>
      </FormControl>
    </form>
  )
}

export default LoginForm
