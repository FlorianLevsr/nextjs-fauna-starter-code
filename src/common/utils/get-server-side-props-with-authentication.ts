import { GetServerSideProps, GetServerSidePropsResult } from "next";
import { createFaunaApolloClient } from "./fauna-apollo-client";
import { CurrentUserData, query } from "../context/AuthContext";
import FaunaTokenManager from "./fauna-token-manager";
import * as cookie from 'cookie';
import { User } from "../types/fauna";

type GetServerSidePropsCallback = (currentUser: User | undefined) => Promise<GetServerSidePropsResult<{ [key: string]: any; }>>;

interface GetServerSidePropsWithAuthenticationOptions {
  callback?: GetServerSidePropsCallback;
  redirectOnAuthenticated?: boolean;
  destination?: string;
}

// TODO: Refactor as own file
const xor = (a: boolean, b: boolean) => Boolean(Number(a) ^ Number(b));

const getServerSidePropsWithAuthentication = (options: GetServerSidePropsWithAuthenticationOptions): GetServerSideProps =>
  async (context) => {
    const { callback, redirectOnAuthenticated, destination } = options;

    if (typeof context.req === 'undefined') {
      throw new Error();
    }
  
    let parsedCookies: Record<string, string>;
    if (typeof context.req.headers.cookie === 'undefined') {
      parsedCookies = {};
    } else {
      parsedCookies = cookie.parse(context.req.headers.cookie);
    }
  
    const faunaTokenManager = new FaunaTokenManager(parsedCookies);
  
    const client = createFaunaApolloClient(faunaTokenManager.get());
    const response = await client.query<CurrentUserData>({ query });

    if (xor(response.data.currentUser === null, redirectOnAuthenticated || false)) {
      return {
        redirect: {
          permanent: false,
          destination: destination || '/login'
        }
      }
    }

    if (typeof callback === 'undefined') {
      return { props: {} };
    }

    const result = await callback(response.data.currentUser);
    return result;
  }

export default getServerSidePropsWithAuthentication;
