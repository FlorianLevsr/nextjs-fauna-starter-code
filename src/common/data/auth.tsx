import React from 'react'
import { gql } from '@apollo/client/core'
import { useMutation, useQuery } from '@apollo/client/react'
import { useRouter } from 'next/router'
import { createContext, FC } from 'react'
import { User } from '../types/fauna'
import { FaunaTokenManager } from '../utils'
import { checkDefined } from '../utils/type-checks'
import { useContext } from 'react'

/**
 * SECTION Interfaces
 */

// ANCHOR Query data structure
export interface CurrentUserData {
  currentUser?: User | null
}

// ANCHOR Login mutation return data structure
interface LoginData {
  loginUser: string
}

// ANCHOR Login input data structure
interface LoginInput {
  username: string
  password: string
}
/**
 * !SECTION
 */

/**
 * SECTION GraphQL queries
 */

// ANCHOR Describe query
export const query = gql`
  query CurrentUserQuery {
    currentUser {
      _id
      username
    }
  }
`

// ANCHOR Describe login query
export const loginQuery = gql`
  mutation LoginUser($username: String!, $password: String!) {
    loginUser(input: { username: $username, password: $password })
  }
`

// ANCHOR Describe logout query
const logoutQuery = gql`
  mutation LogoutUser {
    logoutUser
  }
`
/**
 * !SECTION
 */

/**
 * SECTION Context
 */

// ANCHOR Context value structure
interface AuthContextValue extends CurrentUserData {
  states: {
    mutationLoading: boolean
  }
  actions: {
    login: (username: string, password: string) => void
    logout: () => void
  }
}

// ANCHOR Context creation
export const AuthContext = createContext<AuthContextValue | undefined>(
  undefined
)

// ANCHOR Use Context hook
export const useAuthContext = (): AuthContextValue =>
  checkDefined(
    useContext(AuthContext),
    'AuthContext should not be undefined. Did you forget yo wrap your component inside a Provider?'
  )

// ANCHOR Context provider
export const AuthContextProvider: FC = ({ children }) => {
  /**
   * SECTION Apollo hooks
   */

  // ANCHOR Send request using Apollo client
  const faunaTokenManager = new FaunaTokenManager()
  const { data, refetch } = useQuery<CurrentUserData>(query, {
    errorPolicy: 'all',
  })

  const router = useRouter()

  // ANCHOR Mutation which allows to log in the application
  const [loginMutation, { loading: mutationLoading }] = useMutation<
    LoginData,
    LoginInput
  >(loginQuery, {
    onCompleted: (data) => {
      faunaTokenManager.set(data.loginUser)
      refetch()
    },
  })

  // ANCHOR Mutation which allows to log out from the application
  const [logoutMutation] = useMutation(logoutQuery, {
    onCompleted: () => {
      faunaTokenManager.reset()
      refetch()
    },
  })
  /**
   * !SECTION
   */

  // ANCHOR Pack data and actions to dispatch through components
  let value = {
    states: {
      mutationLoading,
    },
    actions: {
      login: (username: string, password: string) =>
        loginMutation({ variables: { username, password } }),
      logout: () => {
        logoutMutation()
        router.push('/')
      },
    },
  }

  if (typeof data !== 'undefined') {
    value = { ...value, ...data }
  }

  // ANCHOR Template
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
}
