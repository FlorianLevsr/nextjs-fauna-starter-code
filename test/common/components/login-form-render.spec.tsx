import { InMemoryCache } from '@apollo/client'
import React from 'react'
import testRenderer from 'react-test-renderer'
import { LoginForm } from '../../../src/common/components'
import { TestAuth } from '../../utils/auth'

/**
 * SECTION Test suite
 */
describe('Login form', () => {
  let cache: InMemoryCache

  // ANCHOR Unit test - check for the component's HTML rendering
  it('should render correctly', () => {
    // Create a static render
    const tree = testRenderer
      .create(
        <TestAuth cache={cache}>
          <LoginForm
            callback={async () => undefined}
            mutationResult={{ loading: false, called: true }}
          />
        </TestAuth>
      )
      .toJSON()

    // Check the static render against the snapshot
    expect(tree).toMatchSnapshot()
  })
})
