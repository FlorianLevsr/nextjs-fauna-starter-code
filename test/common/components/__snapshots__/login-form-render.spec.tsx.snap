// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form should render correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="chakra-form-control css-1kxonj9"
    role="group"
  >
    <label
      className="chakra-form__label css-uvho5g"
      htmlFor="field-1"
      id="field-1-label"
    >
      Username
      <input
        className="chakra-input css-0"
        disabled={false}
        id="field-1"
        name="username"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        readOnly={false}
        required={false}
        type="text"
      />
    </label>
    <label
      className="chakra-form__label css-uvho5g"
      htmlFor="field-1"
      id="field-1-label"
    >
      Password
      <input
        className="chakra-input css-0"
        disabled={false}
        id="field-1"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        readOnly={false}
        required={false}
        type="password"
      />
    </label>
    <div
      className="css-gmuwbf"
    >
      <button
        className="chakra-button css-1xx8uq7"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </div>
</form>
`;
